### PR TITLE
[Impeller] exploit perfect hash for SamplerDescriptor.

### PR DESCRIPTION
**Pull request by jonahwilliams**
There are only 3 or 4 sampler's active at any given time in a flutter app. rather than store them in a hashmap, just use a vector.

### CODE REVIEW

File: impeller/core/sampler_descriptor.h [hunk 1]
```diff
@@ -30,20 +29,17 @@ struct SamplerDescriptor final : public Comparable<SamplerDescriptor> {
...
-  }
-
-  // Comparable<SamplerDescriptor>
-  bool IsEqual(const SamplerDescriptor& o) const override {
-    return min_filter == o.min_filter && mag_filter == o.mag_filter &&
-           mip_filter == o.mip_filter &&
-           width_address_mode == o.width_address_mode &&
-           height_address_mode == o.height_address_mode &&
-           depth_address_mode == o.depth_address_mode;
+  static uint32_t ToKey(const SamplerDescriptor& d) {
```
Comment (jason-simmons) [inline impeller/core/sampler_descriptor.h:32]: This should return a `uint64_t`

**Review by jason-simmons — COMMENTED** [review]
The crash is happening because the `std::unique_ptr` instances in the `samplers_` vector will be invalidated if the vector is enlarged and its buffer is reallocated.

The `TextureAndSampler` struct used by the render pass is holding a `std::unique_ptr<Sampler>*` to a `unique_ptr` that may become invalidated.

`TextureAndSampler` should instead hold the `Sampler*` raw pointer.

**Review by jonahwilliams — COMMENTED** [review]
Updated this to use the raw_ptr wrapper type

File: impeller/core/sampler_descriptor.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/core/sampler_descriptor.h]: Done

**Review by jason-simmons — APPROVED** [review]

**Pull request merged**